### DISCUSSION TITLE
parameters of e3nn.nn.Gate

### DISCUSSION THREAD

**Discussion by YoavShamir5** [e3nn/e3nn#387] [Q&A]
Hello,

I have read the documentation regarding gates, but still have a few points that are unclear for me:

**1)** Assuming that my input irreps to this gate are "25x0e + 25x1o + 25x2e", what are the constraints/limitations on the values of irreps_scalars, irreps_gates and irreps_gated (while preserving equivariance)? what are the considerations regarding parity? I am mainly struggling with passing non-scalar irreps through the gate.

**2)** What are the considerations regarding parity of these irrep parameters in terms of preserving equivariance?

**3)** for act_scalars and act_gates - what are the considerations for the choice of the torch activation functions, in terms of parity/equivariance?

**4)** finally, somewhat unrelated - if we assume that the output of some layer in my model includes irreps  "25x0e + 25x1o + 25x2e", how can I then create an object that only includes elements of one rotation order, for example - how can I only get the 25 scalar features corresponding to 25x0e output of this layer, discarding the elements of rotation orders 1 and 2?

Thanks for your time and help!
Yoav Shamir

**Comment by YoavShamir5** [comment]
Hello again, would appreciate any input on my queries regarding use cases of gates, thanks in advance!
Yoav Shamir

**Comment by mariogeiger** [comment]
1) irreps_gated = 25x1o+25x2e
You need therefore 50 scalars for irreps_gates, I your input you don't have enough scalars to gate all the non scalars. If you had remaining scalar you have to put them in irreps_scalars

2) the code should handle all the cases. The simplest case is to use even scalars as gates. 

3) When using odd scalars you have to use activation functions with defined parity. 

4) You just have to slice your tensor. (For that, the jax version of e3nn offers better functionalities)

I recommend to test the end-to-end equivariance property of your model

**Reply by VinaySingh561** [reply]
Hello, @mariogeiger could you please confirm whether JAX-based implementations exist for the Compose module of the e3nn.nn.models.gate_points_2101  in the E3NN library? This information would be greatly appreciated

**Reply by mariogeiger** [reply]
You have to reimplement it yourself for your use case. I decided not to provide complete models in e3nn-jax since it always varies for the different applications. But check this example https://github.com/e3nn/e3nn-jax/blob/main/e3nn_jax/experimental/point_convolution.py
It's quite similar

**Comment by YoavShamir5** [comment]
Thanks a lot for these helpful answers. I would appreciate your help with a follow-up - I wonder - can you use a regular TensorProduct (as used in https://docs.e3nn.org/en/stable/api/nn/models/gate_points_2101.html, for example) to to produce, during convolution, enough scalars for irreps_gates, in addition to the regular scalar irreps? for example - if the input to the convolution is 10 scalars, and lmax is 2, can one use a regular TensorProduct (and not a fully connected one) to get as output 30 scalars (10 for irreps_scalars, and 20 for irreps_gates), 10 vectors and 10 tensors of order 2? I am not sure how to produce 30 scalars, and not just 10 scalars. thanks for your help

**Reply by mariogeiger** [reply]
You can multiply by a matrix 30x10 of parameters
